### DISCUSSION TITLE
game version detection is weird

### DISCUSSION THREAD

**Discussion by fayer3** [Kira-NT/mc-publish#94] [Q&A] [answered]
so not sure if that is just an error on my side, but I tried the automatic game version detection, and it didn't seem to work correctly
I was using `Kir-Antipov/mc-publish@v3.2`
here the workflow https://github.com/Vivecraft/VivecraftMod/blob/1.20.1-1.1.0/.github/workflows/publish-on-release.yml#L184-L202
I'll give the fabric versions from the mod json, but forge mostly behaved the same, except for 1.20.x
so for just 1.18.2 it correctly only chose that
```
    "minecraft": "1.18.2",
```
for 1.19.2 it chose 1.19.2, 1.19.3, 1.19.4
```
    "minecraft": "1.19.2",
```
for 1.19.3 it chose 1.19.3, 1.19.4
```
    "minecraft": "1.19.3",
```
for 1.19.4 it worked
```
    "minecraft": "1.19.4",
```
now here comes the weird part on forge it correctly chose the version, although again up to 1.20.2, even when not specified
on fabric though it chose 1.1?
```
    "minecraft": [
      "1.20",
      "1.20.1"
    ],
```
```
    "minecraft": [
      "1.20.2"
    ],
```

I saw that 3.3 adds an `exact` version resolver, would that solve my issue?

and for future stuff, is it possible to let the action run in some kind of dummy state, where it will not publish stuff, but tell what version it got?

**Comment by Kira-NT** [comment, marked as answer]
> I was using `Kir-Antipov/mc-publish@v3.2`

`v3.2` has been out of support since May 18th, 2023. It's not even supposed to work at all anymore, as it requires a framework which is no longer available in the GitHub Actions context, namely Node 12. I assume GitHub now forcefully runs outdated Node 12 actions on Node 16. This might work for simple actions, but will undoubtedly cause `mc-publish` to blow up somewhere down the road in the most unexpected way possible. So, please, do not continue to use it. Update to `v3.3` instead.

With that said, your issues with `mc-publish` are as outdated as `v3.2` is, but I'll still provide some insight into them.

> it didn't seem to work correctly

Actually, it did. It functioned precisely as described in the README at the time. By default, `<=v3.2` interpreted all versions as if they were defined using `>=`. This somewhat finicky implementation was caused by the fact that I didn't really want to deal with Minecraft version ranges when I first started this project, because it would mean processing things like `>=1.RV-Pre1 <=3D Shareware v1.34 || 2point0_purple`. So, I implemented a simple workaround *(which extracts a first Minecraft version from the given string)*, slapped `version-resolver` on top of it, and moved on. I know that this was a terrible design decision (like all quick and dirty solutions are), and I'm glad that it's finally gone.

`>=v3.3` can process Minecraft version ranges and produce results that end users actually expect, instead of merely extracting a single version from the provided range and prefixing it with `>=`.

> on fabric though it chose 1.1?
>
>     "minecraft": [
>       "1.20",
>       "1.20.1"
>     ],

This part was indeed a bug, and even `v3.3` had it for a few days. I wasn't aware that Fabric allowed versions to be specified in an array format. Hence, receiving an array instead of the anticipated string disrupted some parts of my code. While `<=v3.2` defaulted to ">=1.0" (I assume), `v3.3` was better in that it discarded unknown inputs.

As a side note, this bug is already addressed in the latest build of `v3.3`.

> I saw that 3.3 adds an `exact` version resolver, would that solve my issue?

No, `v3.3` doesn't introduce such a feature. Instead, it marks the entire `version-resolver` input as obsolete, as it is no longer relevant and will be removed in `v4.0`.

However, had you used `version-resolver: exact` in `<=v3.2` (where it was available), it would have been beneficial. As mentioned earlier, `<=v3.2` parsed version ranges by pulling a single version from the range and converting it to a `>={version}` SemVer range. Using `version-resolver: exact` would modify this to `={version}`.

> and for future stuff, is it possible to let the action run in some kind of dummy state, where it will not publish stuff, but tell what version it got?

Not at the moment, I'm afraid, but I'm planning to introduce a spec-compliant debug mode in `v4.0`. However, I hope you won't need this, as `>=v3.3` should yield the expected results for you by default, as it interprets the provided version range as is, without any additional magic.

**Reply by fayer3** [reply]
thanks that cleared it up, I'll message back, if I have issues with 3.3

**Reply by fayer3** [reply]
updated to 3.3, fabric versions were fine now
for forge it didn't like parsing single version ranges like this though
```
versionRange = "[1.18.2]"
```
with those it defaulted to 1.0

**Reply by Kira-NT** [reply]
Oh god, they special cased a single version. Of course they did.

Welp, please, open an issue, so I won't forget to add support for this range-which-is-not-actually-a-range syntax.

**Reply by fayer3** [reply]
will do

**Reply by Kira-NT** [reply]
Thanks a lot!
